File: Backtracking/leetcode_46.py

# Problem link ===>>  https://leetcode.com/problems/permutations/description/
# submission link == =>> https://leetcode.com/problems/permutations/submissions/1493387722/

class List:
    def __init__(self):
        self.data = []

    def append(self, item):
        self.data.append(item)

class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:

        def backtrack(start):
            if start == len(nums):
                # Append a copy of the current permutation
                result.append(nums.copy())
                return

            for i in range(start, len(nums)):
                nums[start], nums[i] = nums[i], nums[start]  # Swap elements
                backtrack(start + 1)  # Recursive call
                # Backtrack (undo swap)
                nums[start], nums[i] = nums[i], nums[start]

        result = []
        backtrack(0)
        return result


# Example usage
if __name__ == "__main__":
    solution = Solution()
    nums = [1, 2, 3]
    permutations = solution.permute(nums)

    print("All permutations:")
    for perm in permutations:
        print(perm)

File: Backtracking/leetcode_46.cpp

// Problem link ===>>  https://leetcode.com/problems/permutations/description/


#include <bits/stdc++.h>
#define ll long long
using namespace std;

class Solution
{
public:
    void backtrack(vector<int> &nums, int start, vector<vector<int>> &result)
    {
        if (start == nums.size())
        {
            result.push_back(nums);
            return;
        }
        for (int i = start; i < nums.size(); i++)
        {
            swap(nums[start], nums[i]);
            backtrack(nums, start + 1, result);
            swap(nums[start], nums[i]); // backtrack
        }
    }

    vector<vector<int>> permute(vector<int> &nums)
    {
        vector<vector<int>> result;
        backtrack(nums, 0, result);
        return result;
    }
};

int main()
{
    ios_base::sync_with_stdio(false);
    cout.tie(nullptr);
    cin.tie(nullptr);

    ll t;
    cin >> t;
    while (t--)
    {
        Solution s;
        int n; // Declare n to store the size of the input
        cin >> n;
        vector<int> nums(n); // Initialize nums with size n
        for (int i = 0; i < n; i++)
        {
            cin >> nums[i];
        }

        vector<vector<int>> permutations = s.permute(nums);

        // Print the permutations (important to see the output)
        for (const auto &perm : permutations)
        {
            for (int val : perm)
            {
                cout << val << " ";
            }
            cout << endl;
        }
    }
    return 0;
}